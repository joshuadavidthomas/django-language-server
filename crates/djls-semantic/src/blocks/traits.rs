use djls_templates::{Node, NodeList};

use crate::Db;

/// Semantic model builder that operates on Django template nodelists.
///
/// This trait defines the interface for building semantic models from Django templates.
/// A semantic model is any representation that captures some aspect of the template's
/// meaning - structure, dependencies, types, security properties, etc.
pub trait SemanticModel<'db> {
    /// The semantic model this builder produces
    type Model;

    /// Build the semantic model from a nodelist
    fn model(mut self, db: &'db dyn Db, nodelist: NodeList<'db>) -> Self::Model
    where
        Self: Sized,
    {
        // Default implementation: traverse and observe all nodes
        for node in nodelist.nodelist(db).iter().cloned() {
            self.observe(node);
        }
        self.construct()
    }

    /// Observe a single node during traversal and extract semantic information
    fn observe(&mut self, node: Node<'db>);

    /// Construct the final semantic model from observed semantics
    fn construct(self) -> Self::Model;
}

// Example implementations showing how different semantic models can be built
// from the same Django template source
//
// Example usage:
// ```
// use djls_semantic::blocks::{BlockModelBuilder, SemanticModel};
//
// // Build a block structure model
// let block_tree = BlockModelBuilder::new(db, shapes)
//     .model(db, nodelist);
//
// // Future: Build other semantic models from the same template
// // let inheritance = InheritanceModelBuilder::new(db)
// //     .model(db, nodelist);
// //
// // let dependencies = DependencyModelBuilder::new(db)
// //     .model(db, nodelist);
// ```

/*
// Example future implementations:
/// Builder for extracting template inheritance relationships
pub struct InheritanceTreeBuilder<'db> {
    db: &'db dyn Db,
    extends_from: Option<String>,
    blocks: HashMap<String, BlockInfo>,
}

impl<'db> TemplateTreeBuilder<'db> for InheritanceTreeBuilder<'db> {
    type Output = InheritanceTree;

    fn handle_node(&mut self, node: Node<'db>) {
        if let Node::Tag { name, bits, .. } = node {
            match name.text(self.db) {
                "extends" => {
                    // Extract parent template name
                    if let Some(first_bit) = bits.first() {
                        self.extends_from = Some(first_bit.text(self.db).to_string());
                    }
                }
                "block" => {
                    // Track block definitions
                    if let Some(block_name) = bits.first() {
                        self.blocks.insert(
                            block_name.text(self.db).to_string(),
                            BlockInfo { /* ... */ }
                        );
                    }
                }
                _ => {} // Ignore other tags
            }
        }
    }

    fn finalize(self) -> InheritanceTree {
        InheritanceTree {
            extends_from: self.extends_from,
            blocks: self.blocks,
        }
    }
}

/// Builder for extracting variable dependencies
pub struct VariableDependencyBuilder<'db> {
    db: &'db dyn Db,
    variables: HashSet<String>,
    filters: HashMap<String, Vec<String>>,
}

impl<'db> TemplateTreeBuilder<'db> for VariableDependencyBuilder<'db> {
    type Output = VariableDependencyMap;

    fn handle_node(&mut self, node: Node<'db>) {
        if let Node::Variable { var, filters, .. } = node {
            self.variables.insert(var.text(self.db).to_string());
            // Track filter usage
            for filter in filters {
                self.filters
                    .entry(var.text(self.db).to_string())
                    .or_default()
                    .push(filter.text(self.db).to_string());
            }
        }
    }

    fn finalize(self) -> VariableDependencyMap {
        VariableDependencyMap {
            variables: self.variables,
            filters: self.filters,
        }
    }
}
*/

