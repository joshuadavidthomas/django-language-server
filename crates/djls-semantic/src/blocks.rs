mod builder;
mod grammar;
mod snapshot;
mod tree;

pub(crate) use builder::BlockTreeBuilder;
pub use grammar::TagIndex;
pub(crate) use tree::BlockId;
pub(crate) use tree::BlockNode;
pub(crate) use tree::BlockTree;
pub use tree::BlockTreeInner;
pub(crate) use tree::BranchKind;

use crate::db::Db;
use crate::traits::SemanticModel;

/// Build a block tree from pure data without database
#[doc(hidden)]
pub fn build_block_tree_from_parts(
    specs: &crate::TagSpecs,
    nodes: &[djls_templates::Node],
) -> (BlockTreeInner, Vec<crate::ValidationError>) {
    let index = TagIndex::from_specs(specs);
    let mut builder = BlockTreeBuilder::new(index);
    for node in nodes {
        builder.observe(node.clone());
    }
    builder.construct()
}

// This function is now replaced by the one in queries.rs
// Keep it for backward compatibility but mark as deprecated

#[deprecated(note = "Use queries::build_block_tree instead")]
#[salsa::tracked]
pub fn build_block_tree<'db>(
    db: &'db dyn Db,
    nodelist: djls_templates::NodeList<'db>,
) -> BlockTree<'db> {
    // Forward to the new query
    crate::queries::build_block_tree(db, nodelist)
}