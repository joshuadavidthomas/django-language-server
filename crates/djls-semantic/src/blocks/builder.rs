use djls_source::Span;
use djls_templates::tokens::TagDelimiter;
use djls_templates::Node;

use super::grammar::CloseValidation;
use super::grammar::TagClass;
use super::grammar::TagIndex;
use super::tree::BlockId;
use super::tree::BlockNode;
use super::tree::BlockTree;
use super::tree::BranchKind;
use crate::traits::SemanticModel;
use crate::Db;

#[derive(Debug, Clone)]
enum BlockSemanticOp {
    AddRoot {
        id: BlockId,
    },
    AddBranchNode {
        target: BlockId,
        tag: String,
        marker_span: Span,
        body: BlockId,
        kind: BranchKind,
    },
    AddErrorNode {
        target: BlockId,
        message: String,
        span: Span,
    },
    AddLeafNode {
        target: BlockId,
        label: String,
        span: Span,
    },
    ExtendBlockSpan {
        id: BlockId,
        span: Span,
    },
    FinalizeSpanTo {
        id: BlockId,
        end: u32,
    },
}

pub struct BlockTreeBuilder<'db> {
    db: &'db dyn Db,
    index: &'db TagIndex,
    stack: Vec<TreeFrame>,
    block_allocs: Vec<(Span, Option<BlockId>)>,
    semantic_ops: Vec<BlockSemanticOp>,
}

impl<'db> BlockTreeBuilder<'db> {
    #[allow(dead_code)] // use is gated behind cfg(test) for now
    pub fn new(db: &'db dyn Db, index: &'db TagIndex) -> Self {
        Self {
            db,
            index,
            stack: Vec::new(),
            block_allocs: Vec::new(),
            semantic_ops: Vec::new(),
        }
    }

    /// Allocate a new `BlockId` and track its metadata for later creation
    fn alloc_block_id(&mut self, span: Span, parent: Option<BlockId>) -> BlockId {
        let id = BlockId::new(u32::try_from(self.block_allocs.len()).unwrap_or_default());
        self.block_allocs.push((span, parent));
        id
    }

    /// Apply all semantic operations to build a `BlockTree`
    fn apply_operations(self) -> BlockTree {
        let mut tree = BlockTree::new();

        // Allocate all blocks using metadata
        for (span, parent) in self.block_allocs {
            if let Some(p) = parent {
                tree.blocks_mut().alloc(span, Some(p));
            } else {
                tree.blocks_mut().alloc(span, None);
            }
        }

        for op in self.semantic_ops {
            match op {
                BlockSemanticOp::AddRoot { id } => {
                    tree.roots_mut().push(id);
                }
                BlockSemanticOp::AddBranchNode {
                    target,
                    tag,
                    marker_span,
                    body,
                    kind,
                } => {
                    tree.blocks_mut().push_node(
                        target,
                        BlockNode::Branch {
                            tag,
                            marker_span,
                            body,
                            kind,
                        },
                    );
                }
                BlockSemanticOp::AddLeafNode {
                    target,
                    label,
                    span,
                } => {
                    tree.blocks_mut()
                        .push_node(target, BlockNode::Leaf { label, span });
                }
                BlockSemanticOp::AddErrorNode {
                    target,
                    message,
                    span,
                } => {
                    tree.blocks_mut()
                        .push_node(target, BlockNode::Error { message, span });
                }
                BlockSemanticOp::ExtendBlockSpan { id, span } => {
                    tree.blocks_mut().extend_block(id, span);
                }
                BlockSemanticOp::FinalizeSpanTo { id, end } => {
                    tree.blocks_mut().finalize_block_span(id, end);
                }
            }
        }

        tree
    }

    fn handle_tag(&mut self, name: &String, bits: &Vec<String>, span: Span) {
        let tag_name = name;
        match self.index.classify(tag_name) {
            TagClass::Opener => {
                let parent = get_active_segment(&self.stack);

                let container = self.alloc_block_id(span, parent);
                let segment = self.alloc_block_id(
                    Span::new(span.end().saturating_add(TagDelimiter::LENGTH_U32), 0),
                    Some(container),
                );

                if let Some(parent_id) = parent {
                    // Nested block
                    self.semantic_ops.push(BlockSemanticOp::AddBranchNode {
                        target: parent_id,
                        tag: tag_name.clone(),
                        marker_span: span,
                        body: container,
                        kind: BranchKind::Opener,
                    });
                    self.semantic_ops.push(BlockSemanticOp::AddBranchNode {
                        target: container,
                        tag: tag_name.clone(),
                        marker_span: span,
                        body: segment,
                        kind: BranchKind::Segment,
                    });
                } else {
                    // Root block
                    self.semantic_ops
                        .push(BlockSemanticOp::AddRoot { id: container });
                    self.semantic_ops.push(BlockSemanticOp::AddBranchNode {
                        target: container,
                        tag: tag_name.clone(),
                        marker_span: span,
                        body: segment,
                        kind: BranchKind::Segment,
                    });
                }

                self.stack.push(TreeFrame {
                    opener_name: tag_name.clone(),
                    opener_bits: bits.clone(),
                    opener_span: span,
                    container_body: container,
                    segment_body: segment,
                    parent_body: parent,
                });
            }
            TagClass::Closer { opener_name } => {
                self.close_block(&opener_name, bits, span);
            }
            TagClass::Intermediate { possible_openers } => {
                self.add_intermediate(tag_name, &possible_openers, span);
            }
            TagClass::Unknown => {
                if let Some(segment) = get_active_segment(&self.stack) {
                    self.semantic_ops.push(BlockSemanticOp::AddLeafNode {
                        target: segment,
                        label: tag_name.clone(),
                        span,
                    });
                }
            }
        }
    }

    fn close_block(&mut self, opener_name: &str, closer_bits: &[String], span: Span) {
        if let Some(frame_idx) = find_frame_from_opener(&self.stack, opener_name) {
            // Pop any unclosed blocks above this one
            while self.stack.len() > frame_idx + 1 {
                if let Some(unclosed) = self.stack.pop() {
                    if let Some(parent) = unclosed.parent_body {
                        self.semantic_ops.push(BlockSemanticOp::AddErrorNode {
                            target: parent,
                            message: format!("Unclosed block '{}'", unclosed.opener_name),
                            span: unclosed.opener_span,
                        });
                    }
                    // If no parent, this was a root block that wasn't closed - we could track this separately
                }
            }

            // validate and close
            let frame = self.stack.pop().unwrap();
            match self
                .index
                .validate_close(opener_name, &frame.opener_bits, closer_bits, self.db)
            {
                CloseValidation::Valid => {
                    // Finalize the last segment body to end just before the closer marker
                    let content_end = span.start().saturating_sub(TagDelimiter::LENGTH_U32);
                    self.semantic_ops.push(BlockSemanticOp::FinalizeSpanTo {
                        id: frame.segment_body,
                        end: content_end,
                    });
                    // Extend to include closer
                    self.semantic_ops.push(BlockSemanticOp::ExtendBlockSpan {
                        id: frame.container_body,
                        span,
                    });
                }
                CloseValidation::ArgumentMismatch { arg, expected, got } => {
                    self.semantic_ops.push(BlockSemanticOp::AddErrorNode {
                        target: frame.segment_body,
                        message: format!(
                            "Argument '{arg}' mismatch: expected '{expected}', got '{got}'"
                        ),
                        span,
                    });
                    self.stack.push(frame); // Restore frame
                }
                CloseValidation::MissingRequiredArg { arg, expected } => {
                    self.semantic_ops.push(BlockSemanticOp::AddErrorNode {
                        target: frame.segment_body,
                        message: format!(
                            "Missing required argument '{arg}': expected '{expected}'"
                        ),
                        span,
                    });
                    self.stack.push(frame);
                }
                CloseValidation::UnexpectedArg { arg, got } => {
                    self.semantic_ops.push(BlockSemanticOp::AddErrorNode {
                        target: frame.segment_body,
                        message: format!("Unexpected argument '{arg}' with value '{got}'"),
                        span,
                    });
                    self.stack.push(frame);
                }
                CloseValidation::NotABlock => {
                    // Should not happen as we already classified it
                    if let Some(segment) = get_active_segment(&self.stack) {
                        self.semantic_ops.push(BlockSemanticOp::AddErrorNode {
                            target: segment,
                            message: format!("Internal error: {opener_name} is not a block"),
                            span,
                        });
                    }
                }
            }
        } else if let Some(segment) = get_active_segment(&self.stack) {
            self.semantic_ops.push(BlockSemanticOp::AddErrorNode {
                target: segment,
                message: format!("Unexpected closing tag '{opener_name}'"),
                span,
            });
        }
    }

    fn add_intermediate(&mut self, tag_name: &str, possible_openers: &[String], span: Span) {
        if let Some(frame) = self.stack.last() {
            if possible_openers.contains(&frame.opener_name) {
                // Finalize previous segment body to just before this marker (full start)
                let content_end = span.start().saturating_sub(TagDelimiter::LENGTH_U32);
                let segment_to_finalize = frame.segment_body;
                let container = frame.container_body;

                self.semantic_ops.push(BlockSemanticOp::FinalizeSpanTo {
                    id: segment_to_finalize,
                    end: content_end,
                });

                let body_start = span.end().saturating_add(TagDelimiter::LENGTH_U32);
                let new_segment_id = self.alloc_block_id(Span::new(body_start, 0), Some(container));

                // Add the branch node for the new segment
                self.semantic_ops.push(BlockSemanticOp::AddBranchNode {
                    target: container,
                    tag: tag_name.to_string(),
                    marker_span: span,
                    body: new_segment_id,
                    kind: BranchKind::Segment,
                });

                self.stack.last_mut().unwrap().segment_body = new_segment_id;
            } else {
                let segment = frame.segment_body;
                let opener_name = frame.opener_name.clone();

                self.semantic_ops.push(BlockSemanticOp::AddErrorNode {
                    target: segment,
                    message: format!("'{tag_name}' is not valid in '{opener_name}'"),
                    span,
                });
            }
        } else {
            // Intermediate tag at top level - this is an error
            // Could track this in a separate error list
        }
    }

    fn finish(&mut self) {
        while let Some(frame) = self.stack.pop() {
            if self.index.is_end_optional(&frame.opener_name) {
                // No explicit closer: finalize last segment to end of input (best-effort)
                // We do not know the real end; leave as-is and extend container by opener span only.
                self.semantic_ops.push(BlockSemanticOp::ExtendBlockSpan {
                    id: frame.container_body,
                    span: frame.opener_span,
                });
            } else if let Some(parent) = frame.parent_body {
                self.semantic_ops.push(BlockSemanticOp::AddErrorNode {
                    target: parent,
                    message: format!("Unclosed block '{}'", frame.opener_name),
                    span: frame.opener_span,
                });
            }
        }
    }
}

type TreeStack = Vec<TreeFrame>;

/// Get the currently active segment (the innermost block we're in)
fn get_active_segment(stack: &TreeStack) -> Option<BlockId> {
    stack.last().map(|frame| frame.segment_body)
}

/// Find a frame in the stack by name
fn find_frame_from_opener(stack: &TreeStack, opener_name: &str) -> Option<usize> {
    stack.iter().rposition(|f| f.opener_name == opener_name)
}

struct TreeFrame {
    opener_name: String,
    opener_bits: Vec<String>,
    opener_span: Span,
    container_body: BlockId,
    segment_body: BlockId,
    parent_body: Option<BlockId>, // Can be None for root blocks
}

impl<'db> SemanticModel<'db> for BlockTreeBuilder<'db> {
    type Model = BlockTree;

    fn observe(&mut self, node: Node) {
        match node {
            Node::Tag { name, bits, span } => {
                self.handle_tag(&name, &bits, span);
            }
            Node::Comment { span, .. } => {
                if let Some(parent) = get_active_segment(&self.stack) {
                    self.semantic_ops.push(BlockSemanticOp::AddLeafNode {
                        target: parent,
                        label: "<comment>".into(),
                        span,
                    });
                }
            }
            Node::Variable { span, .. } => {
                if let Some(parent) = get_active_segment(&self.stack) {
                    self.semantic_ops.push(BlockSemanticOp::AddLeafNode {
                        target: parent,
                        label: "<var>".into(),
                        span,
                    });
                }
            }
            Node::Error {
                full_span, error, ..
            } => {
                if let Some(parent) = get_active_segment(&self.stack) {
                    self.semantic_ops.push(BlockSemanticOp::AddLeafNode {
                        target: parent,
                        label: error.to_string(),
                        span: full_span,
                    });
                }
            }
            Node::Text { .. } => {} // Skip text nodes - we only care about Django constructs
        }
    }

    fn construct(mut self) -> Self::Model {
        self.finish();
        self.apply_operations()
    }
}
