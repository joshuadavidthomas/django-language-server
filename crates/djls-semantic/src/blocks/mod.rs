pub mod builder;
pub mod nodes;
pub mod shapes;
pub mod snapshot;
pub mod traits;
pub mod tree;

pub use builder::BlockModelBuilder;
pub use nodes::{BlockId, BlockNode, Blocks, BranchKind, Region};
pub use shapes::{TagClass, TagShape, TagShapes};
pub use traits::SemanticModel;
pub use tree::BlockTree;

