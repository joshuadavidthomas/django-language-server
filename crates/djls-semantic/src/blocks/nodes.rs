use djls_source::Span;
use serde::Serialize;

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize)]
pub struct BlockId(u32);

impl BlockId {
    pub fn new(id: u32) -> Self {
        Self(id)
    }

    pub fn id(self) -> u32 {
        self.0
    }

    pub fn index(self) -> usize {
        self.0 as usize
    }
}

#[derive(Clone, Debug, Default, Serialize)]
pub struct Blocks(Vec<Region>);

impl Blocks {
    pub fn get(&self, id: usize) -> &Region {
        &self.0[id]
    }
}

impl IntoIterator for Blocks {
    type Item = Region;
    type IntoIter = std::vec::IntoIter<Region>;

    fn into_iter(self) -> Self::IntoIter {
        self.0.into_iter()
    }
}

impl<'a> IntoIterator for &'a Blocks {
    type Item = &'a Region;
    type IntoIter = std::slice::Iter<'a, Region>;

    fn into_iter(self) -> Self::IntoIter {
        self.0.iter()
    }
}

impl<'a> IntoIterator for &'a mut Blocks {
    type Item = &'a mut Region;
    type IntoIter = std::slice::IterMut<'a, Region>;

    fn into_iter(self) -> Self::IntoIter {
        self.0.iter_mut()
    }
}

impl Blocks {
    pub fn alloc(&mut self, span: Span, parent: Option<BlockId>) -> BlockId {
        let id = BlockId(self.0.len() as u32);
        self.0.push(Region::new(span, parent));
        id
    }

    pub fn extend_block(&mut self, id: BlockId, span: Span) {
        self.block_mut(id).extend_span(span);
    }

    pub fn push_node(&mut self, target: BlockId, node: BlockNode) {
        let span = node.span();
        self.extend_block(target, span);
        self.block_mut(target).nodes.push(node);
    }

    pub fn block_mut(&mut self, id: BlockId) -> &mut Region {
        let idx = id.index();
        &mut self.0[idx]
    }
}

#[derive(Clone, Debug, Serialize)]
pub struct Region {
    span: Span,
    nodes: Vec<BlockNode>,
    parent: Option<BlockId>,
}

impl Region {
    fn new(span: Span, parent: Option<BlockId>) -> Self {
        Self {
            span,
            nodes: Vec::new(),
            parent,
        }
    }

    pub fn span(&self) -> &Span {
        &self.span
    }

    pub fn set_span(&mut self, span: Span) {
        self.span = span;
    }

    pub fn nodes(&self) -> &Vec<BlockNode> {
        &self.nodes
    }

    fn extend_span(&mut self, span: Span) {
        let opening = self.span.start().saturating_sub(span.start());
        let closing = span.end().saturating_sub(self.span.end());
        self.span = self.span.expand(opening, closing);
    }
}

#[derive(Clone, Debug, Serialize)]
pub enum BranchKind {
    Opener,
    Segment,
}

#[derive(Clone, Debug, Serialize)]
pub enum BlockNode {
    Leaf {
        label: String,
        span: Span,
    },
    Branch {
        tag: String,
        marker_span: Span,
        body: BlockId,
        kind: BranchKind,
    },
    Error {
        message: String,
        span: Span,
    },
}

impl BlockNode {
    fn span(&self) -> Span {
        match self {
            BlockNode::Leaf { span, .. } | BlockNode::Error { span, .. } => *span,
            BlockNode::Branch { marker_span, .. } => *marker_span,
        }
    }
}
