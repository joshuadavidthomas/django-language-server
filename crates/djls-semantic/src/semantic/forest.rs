use djls_source::Span;
use djls_templates::tokens::TagDelimiter;
use djls_templates::Node;
use rustc_hash::FxHashMap;
use serde::Serialize;

use crate::blocks::BlockId;
use crate::blocks::BlockNode;
use crate::blocks::BlockTreeInner;
use crate::blocks::BranchKind;
use crate::traits::SemanticModel;
use crate::Db;

#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize)]
#[doc(hidden)]
pub struct SemanticForestInner {
    pub roots: Vec<SemanticNode>,
}

#[salsa::tracked]
pub struct SemanticForest<'db> {
    #[returns(ref)]
    inner: SemanticForestInner,
}

impl<'db> SemanticForest<'db> {
    pub fn roots(self, db: &'db dyn Db) -> &'db [SemanticNode] {
        &self.inner(db).roots
    }

    pub fn compute_tag_spans(self, db: &'db dyn Db) -> Vec<Span> {
        compute_tag_spans(&self.inner(db).roots)
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize)]
pub enum SemanticNode {
    Tag {
        name: String,
        marker_span: Span,
        arguments: Vec<String>,
        segments: Vec<SemanticSegment>,
    },
    Leaf {
        label: String,
        span: Span,
    },
}

#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize)]
pub struct SemanticSegment {
    pub kind: SegmentKind,
    pub marker_span: Span,
    pub content_span: Span,
    pub arguments: Vec<String>,
    pub children: Vec<SemanticNode>,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize)]
pub enum SegmentKind {
    Main,
    Intermediate { tag: String },
}

type ArgumentIndex = FxHashMap<Span, Vec<String>>;

pub(crate) struct ForestBuilder {
    tree_inner: BlockTreeInner,
    arg_index: ArgumentIndex,
}

impl SemanticModel<'_> for ForestBuilder {
    type Model = SemanticForestInner;

    fn observe(&mut self, node: Node) {
        if let Node::Tag { bits, span, .. } = node {
            self.arg_index.insert(span, bits);
        }
    }

    fn construct(self) -> Self::Model {
        let roots = self.build_roots();
        SemanticForestInner { roots }
    }
}

impl ForestBuilder {
    pub fn new(tree_inner: BlockTreeInner) -> Self {
        Self {
            tree_inner,
            arg_index: ArgumentIndex::default(),
        }
    }

    fn build_roots(&self) -> Vec<SemanticNode> {
        self.tree_inner
            .roots
            .iter()
            .filter_map(|root| self.build_root_tag(*root))
            .collect()
    }

    fn build_root_tag(&self, container_id: BlockId) -> Option<SemanticNode> {
        let container = self.tree_inner.blocks.get(container_id.index());
        for node in container.nodes() {
            if let BlockNode::Branch {
                tag,
                marker_span,
                kind: BranchKind::Segment,
                ..
            } = node
            {
                return Some(self.build_tag_from_container(
                    container_id,
                    tag.clone(),
                    *marker_span,
                ));
            }
        }
        None
    }

    fn build_tag_from_container(
        &self,
        container_id: BlockId,
        tag_name: String,
        opener_marker_span: Span,
    ) -> SemanticNode {
        let segments = self.build_segments(container_id, opener_marker_span);
        let arguments = segments
            .first()
            .map(|segment| segment.arguments.clone())
            .unwrap_or_default();

        SemanticNode::Tag {
            name: tag_name,
            marker_span: opener_marker_span,
            arguments,
            segments,
        }
    }

    fn build_segments(
        &self,
        container_id: BlockId,
        opener_marker_span: Span,
    ) -> Vec<SemanticSegment> {
        let mut segments = Vec::new();
        let container = self.tree_inner.blocks.get(container_id.index());

        for (idx, node) in container.nodes().iter().enumerate() {
            if let BlockNode::Branch {
                tag,
                marker_span,
                body,
                kind: BranchKind::Segment,
            } = node
            {
                let kind = if idx == 0 {
                    SegmentKind::Main
                } else {
                    SegmentKind::Intermediate { tag: tag.clone() }
                };

                let marker = if idx == 0 {
                    opener_marker_span
                } else {
                    *marker_span
                };

                let content_block = self.tree_inner.blocks.get(body.index());
                let arguments = self.lookup_arguments(marker);
                let children = self.build_children(*body);

                segments.push(SemanticSegment {
                    kind,
                    marker_span: marker,
                    content_span: *content_block.span(),
                    arguments,
                    children,
                });
            }
        }

        segments
    }

    fn build_children(&self, block_id: BlockId) -> Vec<SemanticNode> {
        let mut children = Vec::new();
        let block = self.tree_inner.blocks.get(block_id.index());

        for node in block.nodes() {
            match node {
                BlockNode::Leaf { label, span } => {
                    children.push(SemanticNode::Leaf {
                        label: label.clone(),
                        span: *span,
                    });
                }
                BlockNode::Branch {
                    tag,
                    marker_span,
                    body,
                    kind: BranchKind::Opener | BranchKind::Segment,
                } => {
                    children.push(self.build_tag_from_container(*body, tag.clone(), *marker_span));
                }
            }
        }

        children
    }

    fn lookup_arguments(&self, marker_span: Span) -> Vec<String> {
        self.arg_index
            .get(&marker_span)
            .cloned()
            .unwrap_or_default()
    }
}

#[must_use]
pub fn compute_tag_spans(roots: &[SemanticNode]) -> Vec<Span> {
    let mut spans = Vec::new();
    collect_spans_from_roots(roots, &mut spans);
    spans.sort_unstable_by_key(|s| (s.start(), s.end()));
    spans.dedup();
    spans
}

fn collect_spans_from_roots(roots: &[SemanticNode], spans: &mut Vec<Span>) {
    for node in roots {
        collect_spans_from_node(node, spans);
    }
}

fn collect_spans_from_node(node: &SemanticNode, spans: &mut Vec<Span>) {
    match node {
        SemanticNode::Tag {
            marker_span,
            segments,
            ..
        } => {
            spans.push({
                let span = *marker_span;
                span.expand(TagDelimiter::LENGTH_U32, TagDelimiter::LENGTH_U32)
            });
            for segment in segments {
                if let SegmentKind::Intermediate { .. } = segment.kind {
                    spans.push({
                        segment
                            .marker_span
                            .expand(TagDelimiter::LENGTH_U32, TagDelimiter::LENGTH_U32)
                    });
                }
                collect_spans_from_roots(&segment.children, spans);
            }
        }
        SemanticNode::Leaf { .. } => {}
    }
}

#[cfg(test)]
mod tests {
    use std::sync::Arc;
    use std::sync::Mutex;

    use camino::Utf8Path;
    use djls_source::File;
    use djls_templates::parse_template;
    use djls_workspace::FileSystem;
    use djls_workspace::InMemoryFileSystem;
    use insta::assert_yaml_snapshot;

    use super::*;
    use crate::blocks::build_block_tree;
    use crate::build_semantic_forest;
    use crate::templatetags::django_builtin_specs;
    use crate::TagIndex;

    #[salsa::db]
    #[derive(Clone)]
    struct TestDatabase {
        storage: salsa::Storage<Self>,
        fs: Arc<Mutex<InMemoryFileSystem>>,
    }

    impl TestDatabase {
        fn new() -> Self {
            Self {
                storage: salsa::Storage::default(),
                fs: Arc::new(Mutex::new(InMemoryFileSystem::new())),
            }
        }

        fn add_file(&self, path: &str, content: &str) {
            self.fs
                .lock()
                .unwrap()
                .add_file(path.into(), content.to_string());
        }
    }

    #[salsa::db]
    impl salsa::Database for TestDatabase {}

    #[salsa::db]
    impl djls_source::Db for TestDatabase {
        fn read_file_source(&self, path: &Utf8Path) -> std::io::Result<String> {
            self.fs.lock().unwrap().read_to_string(path)
        }
    }

    #[salsa::db]
    impl djls_templates::Db for TestDatabase {}

    #[salsa::db]
    impl crate::Db for TestDatabase {
        fn tag_specs(&self) -> crate::templatetags::TagSpecs {
            django_builtin_specs()
        }

        fn tag_index(&self) -> TagIndex {
            TagIndex::from_specs(&self.tag_specs())
        }
    }

    #[test]
    fn semantic_forest_snapshot() {
        let db = TestDatabase::new();
        let source = r"
{% block header %}
    <h1>Title</h1>
{% endblock header %}

{% if user.is_authenticated %}
    {% for item in items %}
        <li>{{ item }}</li>
    {% empty %}
        <li>No items</li>
    {% endfor %}
{% else %}
    <p>Please log in</p>
{% endif %}
";

        db.add_file("template.html", source);
        let file = File::new(&db, "template.html".into(), 0);
        let nodelist = parse_template(&db, file).expect("should parse");

        let block_tree = build_block_tree(&db, nodelist);
        let forest = build_semantic_forest(&db, block_tree, nodelist);

        assert_yaml_snapshot!(ForestSnapshot::capture(forest, &db));
    }

    #[test]
    fn semantic_forest_intermediate_snapshot() {
        let db = TestDatabase::new();
        let source = r"
{% if user.is_staff %}
    <span>Staff</span>
{% elif user.is_manager %}
    <span>Manager</span>
{% else %}
    <span>Regular</span>
{% endif %}
";

        db.add_file("intermediate.html", source);
        let file = File::new(&db, "intermediate.html".into(), 0);
        let nodelist = parse_template(&db, file).expect("should parse");

        let block_tree = build_block_tree(&db, nodelist);
        let forest = build_semantic_forest(&db, block_tree, nodelist);

        assert_yaml_snapshot!("intermediate", ForestSnapshot::capture(forest, &db));
    }

    #[test]
    fn test_pure_forest_operations() {
        // Create a forest without any database
        let inner = SemanticForestInner {
            roots: vec![SemanticNode::Tag {
                name: "if".into(),
                marker_span: Span::new(0, 10),
                arguments: vec!["user.is_authenticated".into()],
                segments: vec![SemanticSegment {
                    kind: SegmentKind::Main,
                    marker_span: Span::new(0, 10),
                    content_span: Span::new(10, 50),
                    arguments: vec!["user.is_authenticated".into()],
                    children: vec![SemanticNode::Leaf {
                        label: "text".into(),
                        span: Span::new(15, 45),
                    }],
                }],
            }],
        };

        // Test compute_tag_spans without db
        let spans = compute_tag_spans(&inner.roots);
        assert_eq!(spans.len(), 1);
        assert_eq!(spans[0], Span::new(0, 10).expand(2, 2));

        // Test pure validation without db
        let specs = django_builtin_specs();
        let errors = super::super::args::validate_block_tags_pure(&specs, &inner.roots);
        // The if tag should validate successfully with one argument
        assert!(errors.is_empty());
    }

    #[derive(Serialize)]
    struct ForestSnapshot {
        roots: Vec<SemanticNode>,
        tag_spans: Vec<Span>,
    }

    impl ForestSnapshot {
        fn capture(forest: SemanticForest, db: &dyn crate::Db) -> Self {
            let roots = forest.roots(db).to_vec();
            let tag_spans = compute_tag_spans(&roots);

            Self { roots, tag_spans }
        }
    }
}

