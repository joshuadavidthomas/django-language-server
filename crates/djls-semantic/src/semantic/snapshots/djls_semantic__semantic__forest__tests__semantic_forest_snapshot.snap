---
source: crates/djls-semantic/src/semantic/forest.rs
expression: "ForestSnapshot::capture(forest, &db)"
---
roots:
  - Tag:
      name: block
      marker_span:
        start: 3
        length: 14
      arguments:
        - header
      segments:
        - kind: Main
          marker_span:
            start: 3
            length: 14
          content_span:
            start: 3
            length: 36
          arguments:
            - header
          children: []
  - Tag:
      name: if
      marker_span:
        start: 64
        length: 26
      arguments:
        - user.is_authenticated
      segments:
        - kind: Main
          marker_span:
            start: 64
            length: 26
          content_span:
            start: 64
            length: 146
          arguments:
            - user.is_authenticated
          children:
            - Tag:
                name: for
                marker_span:
                  start: 99
                  length: 19
                arguments:
                  - item
                  - in
                  - items
                segments:
                  - kind: Main
                    marker_span:
                      start: 99
                      length: 19
                    content_span:
                      start: 99
                      length: 56
                    arguments:
                      - item
                      - in
                      - items
                    children:
                      - Leaf:
                          label: "<var>"
                          span:
                            start: 135
                            length: 6
                  - kind:
                      Intermediate:
                        tag: empty
                    marker_span:
                      start: 153
                      length: 11
                    content_span:
                      start: 155
                      length: 40
                    arguments: []
                    children: []
        - kind:
            Intermediate:
              tag: else
          marker_span:
            start: 208
            length: 10
          content_span:
            start: 210
            length: 34
          arguments: []
          children: []
tag_spans:
  - start: 1
    length: 18
  - start: 62
    length: 30
  - start: 97
    length: 23
  - start: 151
    length: 15
  - start: 206
    length: 14
